### PR TITLE
[partition-metadata] add_output_metadata refactor

### PR DESCRIPTION
**Pull request by dpeng817**
## Summary & Motivation
I was finding that complex series of ternaries really hard to parse. Most of the complexity is the result of using the same dictionary to represent metadata per output name, and metadata per output name and mapping key. 

Instead, have a top-level accumulator class and add an "OutputMetadataHandle", so that every metadata dictionary has the same scope level. We keep the same checks in place as before, but the logic is much simpler now.

## How I Tested These Changes
Existing tests

### CODE REVIEW

**Comment by dpeng817** [comment]
* **#26562** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26562?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#26561** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26561?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26561?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `master`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Review by OwenKephart — APPROVED** [review]
nice
